### PR TITLE
Add Web Viewer

### PR DESCRIPTION
**Pull request by droogmic**
Implemented in Rust using Yew

Resolves https://github.com/a-b-street/osm2lanes/issues/4

Demo: https://droogmic.github.io/osm2lanes/

### CODE REVIEW

**Comment by enzet** [comment]
Wow! This is just awesome! Thank you for the contribution!

I think, it's better to ask @dabreegster to review since I'm not experienced in Rust.

I tried to draw some icons for different types of lanes. Maybe we can use it in the future.

File: rust/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[workspace]
+members = ["osm2lanes", "osm2lanes-web"]
+
+[profile.release]
+# less code to include into binary
```
Comment (dabreegster) [inline rust/Cargo.toml:5]: I'm curious, are these best practices for yew projects? Performance is definitely not a concern at all for something like this project. I'm fine with all of these, just curious about the origin. FWIW I usually default to

```
[profile.dev.package."*"]
opt-level = 3
```

so even when you compile in debug mode, all the external dependencies are in release mode

File: rust/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+/target
+/dist
```
Comment (dabreegster) [inline rust/.gitignore:2]: 1) The `/dist` here didn't work for me; I wound up with `osm2lanes-web/dist/` showing up as untracked
2) I like specifying the Rust-specific ignores here; would you mind removing `target` from `../.gitignore` and moving `Cargo.lock` here please?

File: rust/Trunk.toml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+[build]
+target = "osm2lanes-web/index.html"
```
Comment (dabreegster) [inline rust/Trunk.toml:2]: I've never used trunk before, but wow, it's fantastic! I want to switch another huge project to using it too

File: rust/osm2lanes-web/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,164 @@
+use std::collections::BTreeMap;
+
+use serde::{Deserialize, Serialize};
+use web_sys::HtmlInputElement;
+use yew::prelude::*;
+
+use osm2lanes::{get_lane_specs_ltr, Config, Direction, DrivingSide, LaneSpec, LaneType};
+
+// Use `wee_alloc` as the global allocator.
```
Comment (dabreegster) [inline rust/osm2lanes-web/src/main.rs:9]: Is this a standard yew thing? Is there a reason to override the allocator for wasm? (I've only ever used `wasm-bindgen` before and never had to touch it)

File: rust/osm2lanes-web/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+type ShouldRender = bool;
+
+struct WebLaneType(LaneType);
+
+impl WebLaneType {
+    /// Represents the lane type as a single character. Always picks one buffer type.
+    pub fn to_char(self) -> char {
+        match self.0 {
+            LaneType::Driving => '🚗',
```
Comment (dabreegster) [inline rust/osm2lanes-web/src/main.rs:21]: Unicode is awesome :D

File: rust/osm2lanes-web/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    let mut map = BTreeMap::new();
+    for line in input.lines() {
+        let (key, val) = line.split_once("=").ok_or("tag must be = separated")?;
+        map.insert(key.to_owned(), val.to_owned());
+    }
+    Ok(map)
+}
+
+const CFG: Config = Config {
+    driving_side: DrivingSide::Right,
```
Comment (dabreegster) [inline rust/osm2lanes-web/src/main.rs:60]: A possible next step (doesn't have to be in this PR) could be to add a checkbox or toggle for this

File: rust/release.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+# To be run by someone with push rights to gh-pages
+# TODO: automate this...
```
Comment (dabreegster) [inline rust/release.sh:4]: I reckon we could add a Github actions workflow this later

**Review by dabreegster — COMMENTED** [review]
Wow, thank you for the PR! This is fantastic -- simple and usable, and makes it much more convenient to just try out the converter so far.

> Helped me see that there is still quite a bit of edge-case handling missing.

Indeed, that's one reason we started this project, to get OSM mappers interested in helping with some of the edge cases. :)  What kind of stuff hasn't been translating well? The biggest missing features I know about are handling per-lane widths.

**Comment by dabreegster** [comment]
Some next step ideas:
- show lane width as well
- Have a text box where you can copy a URL like https://www.openstreetmap.org/way/8363012. We can query Overpass to grab the tags

File: rust/.gitignore (hunk 2 above)
Comment (droogmic) [reply rust/.gitignore:2]: - [ ] gitignore fixes and imprrovements

File: rust/osm2lanes-web/src/main.rs (hunk 4 above)
Comment (droogmic) [reply rust/osm2lanes-web/src/main.rs:9]: > wee_alloc is a tiny allocator that is much smaller than the allocator that is normally used in Rust binaries. Replacing the default allocator with this one will result in smaller Wasm file sizes, at the expense of speed and memory overhead.

> The slower speed and memory overhead are minor in comparison to the size gains made by not including the default allocator. This smaller file size means that your page will load faster, and so it is generally recommended that you use this allocator over the default, unless your app is doing some allocation-heavy work.

Comment (droogmic) [reply rust/osm2lanes-web/src/main.rs:9]: https://yew.rs/docs/next/advanced-topics/optimizations

File: rust/osm2lanes-web/src/main.rs (hunk 6 above)
Comment (droogmic) [reply rust/osm2lanes-web/src/main.rs:60]: yes, a next step please :)

File: rust/release.sh (hunk 7 above)
Comment (droogmic) [reply rust/release.sh:4]: yeah, I tried in the past, and struggled, ill let someone else with more experience with github actions try.

**Comment by droogmic** [comment]
> Some next step ideas:
> 
>     * show lane width as well
> 
>     * Have a text box where you can copy a URL like https://www.openstreetmap.org/way/8363012. We can query Overpass to grab the tags

agreed,

For lane width, I think a canvas renderer would be cool.

For url: something like https://rustwasm.github.io/wasm-bindgen/examples/fetch.html to fetch from overpass.

**Comment by droogmic** [comment]
> Wow, thank you for the PR! This is fantastic -- simple and usable, and makes it much more convenient to just try out the converter so far.
> 
> > Helped me see that there is still quite a bit of edge-case handling missing.
> 
> Indeed, that's one reason we started this project, to get OSM mappers interested in helping with some of the edge cases. :) What kind of stuff hasn't been translating well? The biggest missing features I know about are handling per-lane widths.

just in the example I link, I don't think taxi and bus lanes are correct, and there are more total lanes than lanes says.

**Comment by droogmic** [comment]
I think I am happy with this as an MVP if it works for you, then I can iterate further with smaller PRs.

**Comment by enzet** [comment]
> just in the example I link, I don't think taxi and bus lanes are correct, and there are more total lanes than lanes says.

We decided to use test-driven development. Could you please add test or just write what output we have to generate for this example? I'll try to support it.

**Comment by droogmic** [comment]
> > just in the example I link, I don't think taxi and bus lanes are correct, and there are more total lanes than lanes says.
> 
> We decided to use test-driven development. Could you please add test or just write what output we have to generate for this example? I'll try to support it.

oh sure, I was going to make separate PRs for all of this, dw ;)

**Comment by dabreegster** [comment]
LGTM, thank you for the PR!

**Pull request merged**